### PR TITLE
Convert `apiRoutes` service to ES class

### PR DESCRIPTION
**Pull request by robertknight**
Follow the pattern introduced in https://github.com/hypothesis/client/pull/3296 to convert the API routes service to a class and improve various types along the way.

### CODE REVIEW

File: src/types/api.js [hunk 1]
```diff
@@ -5,6 +5,22 @@
...
+ * @prop {string} method - HTTP method
+ * @prop {string} url - URL template
+ * @prop {string} desc - Description of API route
+ */
+
+/**
+ * Structure of the `links` field of the API index response (`/api`) describing
+ * available API routes.
+ *
+ * @typedef {{ [key: string]: RouteMap|RouteMetadata }} RouteMap
```
Comment (robertknight) [inline src/types/api.js:21]: I tried using `Record<string, RouteMap|RouteMetadata>` here but that turns out not to support the recursive nature of this type.

**Comment by robertknight** [comment]
Code coverage found that one of the tests wasn't executing the intended code paths.

File: src/types/api.js (hunk 1 above)
Comment (esanzgar) [reply src/types/api.js:21]: Understand

**Review by esanzgar — CHANGES_REQUESTED** [review]
All looks good.

Once you make the slow test to not wait I will approve it.

**Comment by esanzgar** [comment]
If you can make the fix for the slow test in a follow up commit, it will help me to review the PR.

File: src/sidebar/services/oauth-auth.js [hunk 2]
```diff
@@ -25,6 +25,9 @@ import { resolve } from '../util/url';
  * Interaction with OAuth endpoints in the annotation service is delegated to
  * the `OAuthClient` class.
  *
+ * @param {Window} $window
+ * @param {import('./api-routes').APIRoutesService} apiRoutes
```
Comment (esanzgar) [inline src/sidebar/services/oauth-auth.js:29]: Thank you for adding these missing types!

**Comment by robertknight** [comment]
I've pushed a commit to fix the slow test.

**Review by esanzgar — APPROVED** [review]
Another options to fix the slow test would have been to use `clock = sinon.useFakeTimers();`.

**Comment by robertknight** [comment]
> Another options to fix the slow test would have been to use clock = sinon.useFakeTimers();.

Yes, that could have worked too. An upside is less code. A downside is that it means the `APIRoutesService` test would have knowledge of implementation details of `retryPromiseOperation`. In this instance I think either approach would be OK, as long as the risks of each are understood.

**Pull request merged**